umad2sim/umad2sim.c

```diff
@@ -116,6 +151,362 @@ static char umad2sim_sysfs_prefix[32];
+static void fd_data_release(struct fd_data_t *fd_data)
+{
+	struct umad_buf_t *ptr;
+
+	while ((ptr = fd_data_dequeue(fd_data)) != NULL)
+		free_umad_buf(ptr);
+
+	mqueue_destroy(fd_data->mqueue);
+	fd_event_destroy(&fd_data->fd_event);
+	free(fd_data);
+}
+
+/* should be called under global lock */
+static int fd_data_enqueue(struct fd_data_t *fd_data, void *data)
+{
+	struct list_elem_t *ptr;
+	int result = 0;
+
+	ptr = mqueue_add_tail(fd_data->mqueue, data);
+	if (!ptr)
+		result = -1;
+	return result;
+}
+
+/* should be called under global lock */
+static struct fd_data_t *get_fd_data(struct umad2sim_dev *dev, unsigned int fd)
+{
+	if (fd < 1024 && fd >= 2048)
+		return NULL;
+
+	if (!dev)
+		return NULL;
+
+	return dev->fds[(fd - 1024) % FD_PER_DEVICE];
+}
+
+/* should be called under global lock */
+static ssize_t fd_data_mqueue_size(struct fd_data_t *fd_data)
+{
+	return mqueue_get_size(fd_data->mqueue);
+}
+
+/* Returns index into dev->fds where new fd_data_t pointer was stored */
+static int get_new_fd(struct umad2sim_dev *dev)
+{
+	int i;
+
+	for (i = 0; i < FD_PER_DEVICE; i++) {
+		if (dev->fds[i] != NULL)
+			continue;
+		dev->fds[i] = fd_data_create();
+		return (dev->fds[i] == NULL) ? -1 : i;
+	}
+	/* all FDs allocated */
+	return -1;
+}
+
+static struct umad2sim_dev *fd_to_dev(unsigned int fd)
+{
+	if (fd >= 2048)
+		return devices[fd - 2048];
+	if (fd >= 1024)
+		return devices[(fd - 1024) / FD_PER_DEVICE];
+
+	return NULL;
+}
+
+static int close_fd(unsigned int fd)
+{
+	struct umad2sim_dev *dev;
+	int i, idx;
+	struct fd_data_t *fd_data;
+
+	if (fd < 1024)
+		return 0;
+	pthread_mutex_lock(&global_devices_mutex);
+	dev = fd_to_dev(fd);
+	if (!dev) {
+		pthread_mutex_unlock(&global_devices_mutex);
+		return 0;
+	}
+	if (fd >= 2048) {
+		sim_client_set_sm(&dev->sim_client, 0);
+		pthread_mutex_unlock(&global_devices_mutex);
+		return 0;
+	}
+
+	fd_data = get_fd_data(dev, fd);
+	if (fd_data)
+		fd_data_release(fd_data);
+
+	for (i = 0; i < 256; i++) {
+		if (dev->agent_fds[i] == fd) {
+			dev->agent_fds[i] = -1;
+			idx = dev->agent_idx[i];
+			dev->agents[idx].id = (uint32_t)(-1);
+			dev->agent_idx[i] = -1;
+			break;
+		}
+	}
+	dev->fds[(fd - 1024) % FD_PER_DEVICE] = NULL;
+	pthread_mutex_unlock(&global_devices_mutex);
+	return 0;
+}
+
+static void *__receiver(void *arg)
+{
+	struct umad2sim_dev *dev = (struct umad2sim_dev *) arg;
+	struct pollfd pfds;
+	struct umad_buf_t *buf;
+	unsigned int mgmt_class;
+	unsigned int fd;
+	struct fd_data_t *fd_data;
+
+	pfds.fd = dev->sim_client.fd_pktin;
+	pfds.events = POLLIN;
+	pfds.revents = 0;
+
+
+	while (1) {
+		if (real_poll(&pfds, 1, -1) < 0) {
+			ERROR("real_poll failure\n");
+			continue;
+		}
+		/* Do real read and post the message to the queue */
+		buf = alloc_umad_buf(sizeof(struct sim_request));
+
+		if (!buf)
+			continue;
+
+		buf->size = umad2sim_read(dev, buf->umad, buf->size, &mgmt_class);
+		pthread_mutex_lock(&global_devices_mutex);
```

Comment:
looks like fd_data_enqueue and fd_data_dequeue can be protected by the fd mutex instead of the global_devices_mutex